llvm/tools/llvm-remote-cache-test/LLVMCASCacheProvider.cpp

```diff
@@ -0,0 +1,231 @@
+// functions as a reference implementation for a \p RemoteCacheProvider.
+//
+//===----------------------------------------------------------------------===//
+
+#include "RemoteCacheProvider.h"
+#include "compilation_caching_kv.grpc.pb.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/CAS/ActionCache.h"
+#include "llvm/CAS/ObjectStore.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/ThreadPool.h"
+#include "llvm/Support/Threading.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace remote_cache_test;
+using namespace compilation_cache_service::cas::v1;
+using namespace compilation_cache_service::keyvalue::v1;
+
+namespace {
+
+class LLVMCASCacheProvider final : public RemoteCacheProvider {
+  ThreadPool Pool{hardware_concurrency()};
+  std::string TempDir;
+  std::unique_ptr<cas::ObjectStore> CAS;
+  std::unique_ptr<cas::ActionCache> ActCache;
+
+public:
+  Error initialize(StringRef CachePath);
+
+  void GetValueAsync(
+      const GetValueRequest &Request,
+      std::function<void(const GetValueResponse &)> Receiver) override;
+  void PutValueAsync(
+      const PutValueRequest &Request,
+      std::function<void(const PutValueResponse &)> Receiver) override;
+
+  void
+  CASLoadAsync(const CASLoadRequest &Request,
+               std::function<void(const CASLoadResponse &)> Receiver) override;
+  void
+  CASSaveAsync(const CASSaveRequest &Request,
+               std::function<void(const CASSaveResponse &)> Receiver) override;
+
+  GetValueResponse GetValue(const GetValueRequest &Request);
+  PutValueResponse PutValue(const PutValueRequest &Request);
+
+  CASLoadResponse CASLoad(const CASLoadRequest &Request);
+  CASSaveResponse CASSave(const CASSaveRequest &Request);
+};
+
+} // namespace
+
+static GetValueResponse GetValueWithError(Error &&E) {
+  GetValueResponse Response;
+  Response.set_outcome(GetValueResponse_Outcome_ERROR);
+  Response.mutable_error()->set_description(toString(std::move(E)));
+  return Response;
+}
+
+static PutValueResponse PutValueWithError(Error &&E) {
+  PutValueResponse Response;
+  Response.mutable_error()->set_description(toString(std::move(E)));
+  return Response;
+}
+
+static CASLoadResponse CASLoadWithError(Error &&E) {
+  CASLoadResponse Response;
+  Response.set_outcome(CASLoadResponse_Outcome_ERROR);
+  Response.mutable_error()->set_description(toString(std::move(E)));
+  return Response;
+}
+
+static CASSaveResponse CASSaveWithError(Error &&E) {
+  CASSaveResponse Response;
+  Response.mutable_error()->set_description(toString(std::move(E)));
+  return Response;
+}
+
+Error LLVMCASCacheProvider::initialize(StringRef CachePath) {
+  TempDir = (Twine(CachePath) + "/tmp").str();
+  if (std::error_code EC = sys::fs::create_directories(TempDir))
+    return createFileError(TempDir, EC);
+
+  auto CAS = cas::createOnDiskCAS(Twine(CachePath) + "/cas");
```

Comment:
Since this is only a test tool, I might just overlay on top of in memory cas since we don't plan to restart the test server and expect data persistency.

Reply:
Actually the lit test (`test/CAS/remote-cache-service.c`) expects being able to connect to a data-persistent service.